subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java

```diff
@@ -319,63 +284,46 @@ DefaultJavaCompileSpec createSpec() {
-        final JavaInstallationMetadata metadata = getToolchain();
+        ForkOptions forkOptions = compileOptions.getForkOptions();
+        JavaInstallationMetadata metadata = getJavaCompiler().get().getMetadata();
         forkOptions.setJavaHome(metadata.getInstallationPath().getAsFile());
-    }
-
-    @Nullable
-    private JavaInstallationMetadata getToolchain() {
-        return javaCompiler.map(JavaCompiler::getMetadata).getOrNull();
+        forkOptions.setExecutable(null);
+        return new DefaultJavaCompileSpecFactory(compileOptions, metadata).create();
     }
 
     private void configureCompatibilityOptions(DefaultJavaCompileSpec spec) {
-        final JavaInstallationMetadata toolchain = getToolchain();
-        if (toolchain != null) {
-            if (compileOptions.getRelease().isPresent()) {
-                spec.setRelease(compileOptions.getRelease().get());
-            } else {
-                boolean isSourceOrTargetConfigured = false;
-                if (super.getSourceCompatibility() != null) {
-                    spec.setSourceCompatibility(getSourceCompatibility());
-                    isSourceOrTargetConfigured = true;
-                }
-                if (super.getTargetCompatibility() != null) {
-                    spec.setTargetCompatibility(getTargetCompatibility());
-                    isSourceOrTargetConfigured = true;
-                }
-                if (!isSourceOrTargetConfigured) {
-                    JavaLanguageVersion languageVersion = toolchain.getLanguageVersion();
-                    if (languageVersion.canCompileOrRun(10)) {
-                        spec.setRelease(languageVersion.asInt());
-                    } else {
-                        String version = languageVersion.toString();
-                        spec.setSourceCompatibility(version);
-                        spec.setTargetCompatibility(version);
-                    }
-                }
-            }
-        } else if (compileOptions.getRelease().isPresent()) {
+        JavaInstallationMetadata toolchain = getJavaCompiler().get().getMetadata();
+        if (compileOptions.getRelease().isPresent()) {
             spec.setRelease(compileOptions.getRelease().get());
         } else {
-            spec.setTargetCompatibility(getTargetCompatibility());
-            spec.setSourceCompatibility(getSourceCompatibility());
+            boolean isSourceOrTargetConfigured = false;
+            if (getSourceCompatibility() != null) {
+                spec.setSourceCompatibility(getSourceCompatibility());
+                isSourceOrTargetConfigured = true;
+            }
+            if (getTargetCompatibility() != null) {
+                spec.setTargetCompatibility(getTargetCompatibility());
+                isSourceOrTargetConfigured = true;
+            }
+            if (!isSourceOrTargetConfigured) {
+                JavaLanguageVersion languageVersion = toolchain.getLanguageVersion();
```

Comment:
This is change, since we'd always set the release/target/source compatibility now. I don't think it is a problem, though, since I don't think setting release to the same version as the toolchain doesn't have any effect. I actually don't understand why we set it here when we can. Maybe I am missing something.